### DISCUSSION TITLE
gigadevice gd32 spi callback error

### DISCUSSION THREAD

**Discussion by lmikl** [zephyrproject-rtos/zephyr#55647] [Q&A] [answered]
hello,

I am currently working with a gd32f427 and I think there is an error in the spi/gd32 library.

I always have the callback function which is called 2 times for each transmission frame on the spi.
I use the spi_transceive_cb function.
I debugged the problem and the spi_gd32_isr function is called 2 times for each transmission frame. but the bits RBNEIE, TBEIE and ERRIE are clear at the first call.

If someone can help me to solve this problem.
thanks in advance.

**Comment by lmikl** [comment]
I have more details about the bug.

If I use the "spi_transceive_cb" function, I always have one more byte sent on the spi bus.
If I send 8 bytes with this function. I always have 9 bytes sent on the spi bus. I checked with an oscilloscope.

But if I use the "spi_transceive_dt" function, it's fine. if I send 8 bytes, I have 8 bytes on the spi bus.

If someone can be checked this bug and solve it.

Thanks

**Reply by lmikl** [reply]
update:

the problem appears when I enable the interrupt for the SPI (CONFIG_SPI_GD32_INTERRUPT).
An additional byte is sent for each complete frame

**Comment by cameled** [comment]
Hi @lmikl, thanks for you feedback.

There have similar issue in https://github.com/zephyrproject-rtos/zephyr/pull/47499#discussion_r928278905.

Since `spi_transceive_dt` api is OK, can you leave the `callback` function empty in `spi_transceive_cb` api? I think this is one different between them.

**Reply by lmikl** [reply]
thanks for the similar issue.
I works with zephyr 3.3.0 and a GD32F427V-START

the problem appears when the spi interrupt is enabled whatever function i use.

**Reply by cameled** [reply]
I have run `samples/drivers/spi_flash` sample on gd32f470i_eval last night. With `CONFIG_SPI_GD32_INTERRUPT` enabled, this sample works fine. The `samples/drivers/spi_flash`  sample based on `spi_nor` flash driver, which use `spi_transceive_dt`.

**Reply by lmikl** [reply]
it all depends on the component that is used.

I wrote a driver to manage the ADIN2111 component. And this one doesn't like that there is an extra byte in the SPI frames before the chip select is disabled.

Maybe the nor flash you are using is not affected by this extra byte.

The only way to see it is to use an oscilloscope and spy on the spi bus.

**Reply by cameled** [reply]
The total length for a spi transceive, is the max length of `rx` buffers or `tx` buffers.

Can you check your `rx` buffers length? or does the extra byte is 0?

**Reply by cameled** [reply]
I will use logic analysizer to check bus conditions tonight.

**Reply by lmikl** [reply]
![spi with interrupt](https://user-images.githubusercontent.com/127130921/229748142-9d6818e4-e9fa-4d22-82c1-daf87f738205.PNG)
![spi without interrupt](https://user-images.githubusercontent.com/127130921/229748155-1b7e479d-a44b-4acf-a894-e30bb3e3b3a5.PNG)
i join 2 screenshoot.
One without interrupt and one with interrupt. You can see the extra byte.
there is the chip select and the clk.
i use the spi_transceive_dt function.

**Reply by cameled** [reply, marked as answer]
Please test with below patch, transceive will be ok.
```
diff --git a/drivers/spi/spi_gd32.c b/drivers/spi/spi_gd32.c
index 37c705ed30..4d705f9037 100644
--- a/drivers/spi/spi_gd32.c
+++ b/drivers/spi/spi_gd32.c
@@ -471,17 +471,19 @@ static void spi_gd32_isr(struct device *dev)
 	struct spi_gd32_data *data = dev->data;
 	int err = 0;
 
-	if ((SPI_STAT(cfg->reg) & SPI_GD32_ERR_MASK) != 0) {
-		err = spi_gd32_get_err(cfg);
-	} else {
+	err = spi_gd32_get_err(cfg);
+	if (err) {
+		spi_gd32_complete(dev, err);
+		return;
+	}
+
+	if (spi_gd32_transfer_ongoing(data)) {
 		err = spi_gd32_frame_exchange(dev);
 	}
 
 	if (err || !spi_gd32_transfer_ongoing(data)) {
 		spi_gd32_complete(dev, err);
 	}
-
-	SPI_STAT(cfg->reg) = 0;
 }
 
 #endif /* SPI_GD32_INTERRUPT */
```

**Reply by lmikl** [reply]
Hi,

I tested your solution and it works fine.
Thank you very much for your response.

I have another question, do you take care of the gigadevice part for zephyr? because I need to use some features of the gigadevice which are not present on zephyr

**Reply by cameled** [reply]
There have a RFC for [Gigadevice Vendor Support](https://github.com/zephyrproject-rtos/zephyr/issues/38657). Currently gd32  supported by community.
